### DISCUSSION TITLE
ESP32 compile error (used to work!)

### DISCUSSION THREAD

**Discussion by mleyden** [zephyrproject-rtos/zephyr#71293] [General]
Hi, I was able to compile for my esp32_devkitc_wroom board without error for some time... I updated hal_espressif and now I've broken it. Any ideas what has gone wrong? The errors suggest compiling with -mlongcalls, which I have tried with zephyr_library_compile_options(-mlongcalls) but doesn't work. 

Get the following:

-- west build: generating a build system
cmake version 3.22.1 is OK; minimum version is 3.13.1
Running CMake: /usr/bin/cmake -DWEST_PYTHON=/usr/bin/python3 -B/workspaces/<MyProject>/build -GNinja -S/workspaces/<MyProject>/app
-- The C compiler identification is GNU 11.4.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Application: /workspaces/<MyProject>/app
-- CMake version: 3.22.1
-- Found Python3: /usr/bin/python3 (found suitable version "3.10.12", minimum required is "3.8") found components: Interpreter 
-- Cache files will be written to: /home/user/.cache/zephyr
-- Zephyr version: 3.5.0 (/workspaces/<MyProject>/zephyr)
-- Found west (found suitable version "1.2.0", minimum required is "0.14.0")
-- Board: esp32_devkitc_wroom
-- Found host-tools: zephyr 0.16.5 (/opt/toolchains/zephyr-sdk-0.16.5-1)
-- Found toolchain: zephyr 0.16.5 (/opt/toolchains/zephyr-sdk-0.16.5-1)
-- Found Dtc: /opt/toolchains/zephyr-sdk-0.16.5-1/sysroots/x86_64-pokysdk-linux/usr/bin/dtc (found suitable version "1.6.0", minimum required is "1.4.6") 
-- Found BOARD.dts: /workspaces/<MyProject>/zephyr/boards/xtensa/esp32_devkitc_wroom/esp32_devkitc_wroom.dts
-- Found devicetree overlay: /workspaces/<MyProject>/app/boards/esp32_devkitc_wroom.overlay
-- Generated zephyr.dts: /workspaces/<MyProject>/build/zephyr/zephyr.dts
-- Generated devicetree_generated.h: /workspaces/<MyProject>/build/zephyr/include/generated/devicetree_generated.h
-- Including generated dts.cmake file: /workspaces/<MyProject>/build/zephyr/dts.cmake
Parsing /workspaces/<MyProject>/zephyr/Kconfig
Loaded configuration '/workspaces/<MyProject>/zephyr/boards/xtensa/esp32_devkitc_wroom/esp32_devkitc_wroom_defconfig'
Merged configuration '/workspaces/<MyProject>/app/prj.conf'
Configuration saved to '/workspaces/<MyProject>/build/zephyr/.config'
Kconfig header saved to '/workspaces/<MyProject>/build/zephyr/include/generated/autoconf.h'
-- Found GnuLd: /opt/toolchains/zephyr-sdk-0.16.5-1/xtensa-espressif_esp32_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32_zephyr-elf/bin/ld.bfd (found version "2.38") 
-- The CXX compiler identification is GNU 12.2.0
-- The ASM compiler identification is GNU
-- Found assembler: /opt/toolchains/zephyr-sdk-0.16.5-1/xtensa-espressif_esp32_zephyr-elf/bin/xtensa-espressif_esp32_zephyr-elf-gcc
-- Configuring done
-- Generating done
-- Build files have been written to: /workspaces/<MyProject>/build
-- west build: building application
cmake version 3.22.1 is OK; minimum version is 3.13.1

...

[156/161] : && ccache /opt/toolchains/zephyr-sdk-0.16.5-1/xtensa-espressif_esp32_zephyr-elf/bin/xtensa-espressif_esp32_zephyr-elf-gcc  -gdwarf-4 zephyr/CMakeFiles/zephyr_pre0.dir/misc/empty_file.c.o -o zephyr/zephyr_pre0.elf  zephyr/CMakeFiles/offsets.dir/./arch/xtensa/core/offsets/offsets.c.o  -fuse-ld=bfd  -T  zephyr/linker_zephyr_pre0.cmd  -Wl,-Map=/workspaces/<MyProject>/build/zephyr/zephyr_pre0.map  -Wl,--whole-archive  app/libapp.a  zephyr/libzephyr.a  zephyr/arch/common/libarch__common.a  zephyr/arch/arch/xtensa/core/libarch__xtensa__core.a  zephyr/lib/libc/picolibc/liblib__libc__picolibc.a  zephyr/lib/libc/common/liblib__libc__common.a  zephyr/subsys/random/libsubsys__random.a  zephyr/drivers/interrupt_controller/libdrivers__interrupt_controller.a  zephyr/drivers/clock_control/libdrivers__clock_control.a  zephyr/drivers/console/libdrivers__console.a  zephyr/drivers/entropy/libdrivers__entropy.a  zephyr/drivers/gpio/libdrivers__gpio.a  zephyr/drivers/pinctrl/libdrivers__pinctrl.a  zephyr/drivers/serial/libdrivers__serial.a  zephyr/drivers/timer/libdrivers__timer.a  -Wl,--no-whole-archive  zephyr/kernel/libkernel.a  -L"/opt/toolchains/zephyr-sdk-0.16.5-1/xtensa-espressif_esp32_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32_zephyr-elf/12.2.0"  -L/workspaces/<MyProject>/build/zephyr  -lgcc  zephyr/arch/common/libisr_tables.a  -lgcc  -Wl,--gc-sections  -Wl,--build-id=none  -Wl,--sort-common=descending  -Wl,--sort-section=alignment  -Wl,-u,_OffsetAbsSyms  -Wl,-u,_ConfigAbsSyms  -nostdlib  -static  -Wl,-X  -Wl,--orphan-handling=warn  -Wl,-no-pie  -DPICOLIBC_INTEGER_PRINTF_SCANF  -T/workspaces/<MyProject>/modules/hal/espressif/zephyr/esp32/src/linker/esp32.rom.alias.ld  -T/workspaces/<MyProject>/modules/hal/espressif/zephyr/esp32/../../components/esp_rom/esp32/ld/esp32.rom.ld  -T/workspaces/<MyProject>/modules/hal/espressif/zephyr/esp32/../../components/esp_rom/esp32/ld/esp32.rom.api.ld  -T/workspaces/<MyProject>/modules/hal/espressif/zephyr/esp32/../../components/esp_rom/esp32/ld/esp32.rom.libgcc.ld  -T/workspaces/<MyProject>/modules/hal/espressif/zephyr/esp32/../../components/esp_rom/esp32/ld/esp32.rom.newlib-data.ld  -T/workspaces/<MyProject>/modules/hal/espressif/zephyr/esp32/../../components/esp_rom/esp32/ld/esp32.rom.newlib-funcs.ld  -T/workspaces/<MyProject>/modules/hal/espressif/zephyr/esp32/../../components/soc/esp32/ld/esp32.peripherals.ld  -Wl,--wrap=k_calloc  -Wl,--wrap=k_malloc  --specs=picolibc.specs  -lc  -lgcc && cd /workspaces/<MyProject>/build/zephyr && /usr/bin/cmake -E true
FAILED: zephyr/zephyr_pre0.elf zephyr/zephyr_pre0.map /workspaces/<MyProject>/build/zephyr/zephyr_pre0.map 
: && ccache /opt/toolchains/zephyr-sdk-0.16.5-1/xtensa-espressif_esp32_zephyr-elf/bin/xtensa-espressif_esp32_zephyr-elf-gcc  -gdwarf-4 zephyr/CMakeFiles/zephyr_pre0.dir/misc/empty_file.c.o -o zephyr/zephyr_pre0.elf  zephyr/CMakeFiles/offsets.dir/./arch/xtensa/core/offsets/offsets.c.o  -fuse-ld=bfd  -T  zephyr/linker_zephyr_pre0.cmd  -Wl,-Map=/workspaces/<MyProject>/build/zephyr/zephyr_pre0.map  -Wl,--whole-archive  app/libapp.a  zephyr/libzephyr.a  zephyr/arch/common/libarch__common.a  zephyr/arch/arch/xtensa/core/libarch__xtensa__core.a  zephyr/lib/libc/picolibc/liblib__libc__picolibc.a  zephyr/lib/libc/common/liblib__libc__common.a  zephyr/subsys/random/libsubsys__random.a  zephyr/drivers/interrupt_controller/libdrivers__interrupt_controller.a  zephyr/drivers/clock_control/libdrivers__clock_control.a  zephyr/drivers/console/libdrivers__console.a  zephyr/drivers/entropy/libdrivers__entropy.a  zephyr/drivers/gpio/libdrivers__gpio.a  zephyr/drivers/pinctrl/libdrivers__pinctrl.a  zephyr/drivers/serial/libdrivers__serial.a  zephyr/drivers/timer/libdrivers__timer.a  -Wl,--no-whole-archive  zephyr/kernel/libkernel.a  -L"/opt/toolchains/zephyr-sdk-0.16.5-1/xtensa-espressif_esp32_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32_zephyr-elf/12.2.0"  -L/workspaces/<MyProject>/build/zephyr  -lgcc  zephyr/arch/common/libisr_tables.a  -lgcc  -Wl,--gc-sections  -Wl,--build-id=none  -Wl,--sort-common=descending  -Wl,--sort-section=alignment  -Wl,-u,_OffsetAbsSyms  -Wl,-u,_ConfigAbsSyms  -nostdlib  -static  -Wl,-X  -Wl,--orphan-handling=warn  -Wl,-no-pie  -DPICOLIBC_INTEGER_PRINTF_SCANF  -T/workspaces/<MyProject>/modules/hal/espressif/zephyr/esp32/src/linker/esp32.rom.alias.ld  -T/workspaces/<MyProject>/modules/hal/espressif/zephyr/esp32/../../components/esp_rom/esp32/ld/esp32.rom.ld  -T/workspaces/<MyProject>/modules/hal/espressif/zephyr/esp32/../../components/esp_rom/esp32/ld/esp32.rom.api.ld  -T/workspaces/<MyProject>/modules/hal/espressif/zephyr/esp32/../../components/esp_rom/esp32/ld/esp32.rom.libgcc.ld  -T/workspaces/<MyProject>/modules/hal/espressif/zephyr/esp32/../../components/esp_rom/esp32/ld/esp32.rom.newlib-data.ld  -T/workspaces/<MyProject>/modules/hal/espressif/zephyr/esp32/../../components/esp_rom/esp32/ld/esp32.rom.newlib-funcs.ld  -T/workspaces/<MyProject>/modules/hal/espressif/zephyr/esp32/../../components/soc/esp32/ld/esp32.peripherals.ld  -Wl,--wrap=k_calloc  -Wl,--wrap=k_malloc  --specs=picolibc.specs  -lc  -lgcc && cd /workspaces/<MyProject>/build/zephyr && /usr/bin/cmake -E true
zephyr/drivers/gpio/libdrivers__gpio.a(gpio_esp32.c.o): in function `gpio_esp32_config':
/workspaces/<MyProject>/zephyr/drivers/gpio/gpio_esp32.c:239:(.text.gpio_esp32_config+0x23f): dangerous relocation: call8: call target out of range: esp_rom_gpio_matrix_out

**Comment by rftafas** [comment]
Mandatory sanity question: by _"I updated hal_espressif"_ you mean you updated everything, Zephyr, HAL, binary blobs and etc. Right?

**Reply by rftafas** [reply]
I also suggest a pristine build after update. https://docs.zephyrproject.org/latest/develop/west/build-flash-debug.html#pristine-builds

**Reply by mleyden** [reply]
> Mandatory sanity question: by _"I updated hal_espressif"_ you mean you updated everything, Zephyr, HAL, binary blobs and etc. Right?

Yep. >west update

I just deleted the /build directory... I see the -p flag is more elegant!

I mentioned hal_espressif as the only thing I could think that changed... as zephyr is fixed to v3.5.0

**Reply by rftafas** [reply]
You need to use Zephyr head - not even 3.6 - in order to use Hal 5.1. For Zephyr 3.5 you should use Hal 4.x. If for whatever reason Zephyr 3.5 called for hal5.1, open a bug report.

**Comment by mleyden** [comment]
Thanks for the help. I eventually tracked down the error. It was in my CMakeLists.txt...
This fails:
cmake_minimum_required(VERSION 3.20.0)
project(app LANGUAGES C)
set(BOARD esp_wrover_kit)
find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
target_sources(app PRIVATE src/main.c)

But this works:

cmake_minimum_required(VERSION 3.20.0)
set(BOARD esp_wrover_kit)
find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
project(app LANGUAGES C)
target_sources(app PRIVATE src/main.c)

So the order of project(app LANGUAGES C) is important. According to https://cmake.org/cmake/help/latest/command/project.html cmake should throw a warning if the order is incorrect, however I don't see it...

Perhaps someone with better cmake experience can shed some light?

Thanks, Mark

**Reply by rftafas** [reply]
Then, as I understand it, you are using plain 3.5.x with the proper hal_espressif (4.x), no tinkering involved on hal version.
